### PR TITLE
Fix/validate ctx

### PR DESCRIPTION
**Pull request by dirkmc**
Pass entire record to validation functions (not just key and value)

### CODE REVIEW

File: validation.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ var log = logging.Logger("routing/record")
 
 // ValidatorFunc is a function that is called to validate a given
 // type of DHTRecord.
-type ValidatorFunc func(string, []byte) error
+type ValidatorFunc func(*pb.Record) error
```
Comment (Stebalien) [inline validation.go]: As this is a public API, I'd prefer to define a new Record struct:

```go
type Record struct {
  Key string
  Value []byte
  Author peer.ID // If and only if signed. Otherwise, just ignore the author (can't trust it).
}
```

It also prevents the validator from looking at anything it shouldn't.

**Comment by dirkmc** [comment]
@Stebalien is this more like what you were imagining?

File: validation.go [hunk 2]
```diff
@@ -53,7 +73,21 @@ func (v Validator) VerifyRecord(r *pb.Record) error {
 		return ErrInvalidRecordType
 	}
 
-	return val.Func(r.GetKey(), r.GetValue())
+	var author *peer.ID
+	if len(r.GetSignature()) > 0 {
+		pid, err := peer.IDFromString(r.GetAuthor())
+		if err != nil {
+fmt.Printf("Could not parse author to peer ID: %s\n", r.GetAuthor())
```
Comment (vyzo) [inline validation.go]: uhm, this seems forgotten from debugging?

File: validation.go [hunk 3]
```diff
@@ -53,7 +73,21 @@ func (v Validator) VerifyRecord(r *pb.Record) error {
...
+			return ErrInvalidRecordType
+		}
+		author = &pid
+	}
+	vr := &ValidationRecord{
```
Comment (vyzo) [inline validation.go]: so how are we going to validate the signature if we don't pass it?

Comment (vyzo) [reply validation.go]: Or is this something we do at a different layer?

File: validation.go (hunk 2 above)
Comment (dirkmc) [reply validation.go]: Oops good catch :)

**Comment by vyzo** [comment]
I really think we should pass the pb record in its entirety to the validator, as was the original proposal.
@Stebalien what's the rationale for restricting access to it? the validator should be given full access imo.

File: validation.go (hunk 3 above)
Comment (dirkmc) [reply validation.go]: This is where the DHT checks it:
https://github.com/libp2p/go-libp2p-kad-dht/blob/5a05448e7cf4003ea56a864c40fb371195adf6e1/records.go#L129

Comment (vyzo) [reply validation.go]: ok, so the signature is considered validated at this layer.

File: validation.go [hunk 4]
```diff
@@ -27,6 +24,29 @@ var ErrBadRecord = errors.New("bad dht record")
 // is not found in the Validator map of the DHT.
 var ErrInvalidRecordType = errors.New("invalid record keytype")
 
+type ValidationRecord struct {
+  key string
+  value []byte
+  author *peer.ID
```
Comment (vyzo) [inline validation.go]: it's a little odd to have a pointer to what is a typedef to a string.
Perhaps we should just compare with the empty string instead of nil, and then we won't have to dereference a pointer.

Comment (dirkmc) [reply validation.go]: I agree it is a little strange. I was trying to decide what was stranger between that and `peer.ID("")`

Comment (vyzo) [reply validation.go]: I think we can just compare with the string literal `""`, without the explicit casting, and it should work.

**Comment by dirkmc** [comment]
@vyzo like this?

**Comment by vyzo** [comment]
@dirkmc yeah looks better. I think it looks good, let's wait for @whyrusleeping and @Stebalien for final merge.

File: validation.go [hunk 5]
```diff
@@ -53,7 +74,21 @@ func (v Validator) VerifyRecord(r *pb.Record) error {
 		return ErrInvalidRecordType
 	}
 
-	return val.Func(r.GetKey(), r.GetValue())
+	author := peer.ID("")
+	if len(r.GetSignature()) > 0 {
```
Comment (vyzo) [inline validation.go]: let's add a comment here that the signature has been validated by the DHT; it's critical to be clear about this imo.

Comment (dirkmc) [reply validation.go]: @vyzo Yes I agree it's hard to follow where all the pieces fit together. In my opinion it would be better for signature validation to happen inside this module, perhaps by passing a public key into the verify function. But it would need some rearchitecting because the signature is optional

Comment (vyzo) [reply validation.go]: yeah, no need to complicate. A comment (and perhaps a TODO) will do for now.

**Comment by dirkmc** [comment]
It looks like the build is failing because of a mismatch between a gx and github version of the go-libp2p-crypto lib. Is there some step I'm missing to fix those errors when running tests?

**Comment by vyzo** [comment]
Normally we `gx-go rw` before running tests, but this should be happening automagically for the tests in ci.
I am not a gx guru, so it's hard to tell what's going wrong here -- @Stebalien?

**Comment by vyzo** [comment]
Oh, seems like you added a dependency to `go-libp2p-peer`? That should be added to `package.json` with `gx import`.

**Comment by dirkmc** [comment]
@vyzo Great that fixed it, thank you

**Comment by whyrusleeping** [comment]
Thanks for this @dirkmc :)

**Comment by Stebalien** [comment]
> what's the rationale for restricting access to it

I forgot that this package actually exposes `pb.Record` type.

Motivation:

1. Nicer types (e.g., typed author). Working with types generated by the protobuf package isn't exactly fun.
2. Impossible to do stupid things like trust the author field when there's no valid signature.
3. Gets rid of any other unrelated fields (e.g., `TimeReceived` which only exists for local record keeping).

However, that would require a separate refactor and would, unfortunately, cause problems around removing the author field.

**Comment by Stebalien** [comment]
Actually, I take that back. We should *not* be preserving the author for unsigned records.
